test/lit/passes/gufa-tnh.wast

```diff
@@ -0,0 +1,1905 @@
+      (local.get $array.fill)
+      (i32.const 7)
+      (i32.const 8)
+      (i32.const 9)
+    )
+    (drop
+      (ref.test $A
+        (local.get $ref.test)
+      )
+    )
+  )
+
+  ;; CHECK:      (func $caller (type $anyref_=>_none) (param $any anyref)
+  ;; CHECK-NEXT:  (call $called
+  ;; CHECK-NEXT:   (ref.cast $A
+  ;; CHECK-NEXT:    (local.get $any)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (ref.cast $A
+  ;; CHECK-NEXT:    (local.get $any)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (ref.cast $B
+  ;; CHECK-NEXT:    (local.get $any)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (ref.cast $B
+  ;; CHECK-NEXT:    (local.get $any)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (ref.cast $B
+  ;; CHECK-NEXT:    (local.get $any)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (ref.cast $B
+  ;; CHECK-NEXT:    (local.get $any)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (ref.cast $B
+  ;; CHECK-NEXT:    (local.get $any)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (ref.cast $B
+  ;; CHECK-NEXT:    (local.get $any)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (ref.cast null $A
+  ;; CHECK-NEXT:    (local.get $any)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $caller (export "out") (param $any anyref)
+    ;; All these casts will be refined to non-nullable, aside from the last
+    ;; param which is but a ref.test.
+    (call $called
+      (ref.cast null $A
+        (local.get $any)
+      )
+      (ref.cast null $A
+        (local.get $any)
+      )
+      (ref.cast null $B
+        (local.get $any)
+      )
+      (ref.cast null $B
+        (local.get $any)
+      )
+      (ref.cast null $B
+        (local.get $any)
+      )
+      (ref.cast null $B
+        (local.get $any)
+      )
+      (ref.cast null $B
+        (local.get $any)
+      )
+      (ref.cast null $B
+        (local.get $any)
+      )
+      (ref.cast null $A
+        (local.get $any)
+      )
+    )
+  )
+)
+
+;; A CallRef with no non-trapping targets must be unreachable if traps never
+;; happen. However, this requires closed world, so we do nothing here. (This
+;; test is mirrored in gufa-tnh-closed, where closed world *is* enabled.)
+(module
+  ;; CHECK:      (type $A (func))
+  (type $A (func))
+
+  ;; CHECK:      (type $funcref_=>_none (func (param funcref)))
+
+  ;; CHECK:      (export "out" (func $caller))
+
+  ;; CHECK:      (func $impossible (type $A)
+  ;; CHECK-NEXT:  (unreachable)
+  ;; CHECK-NEXT: )
+  (func $impossible (type $A)
+    ;; This cannot be called if traps never happen.
+    (unreachable)
+  )
+
+  ;; CHECK:      (func $caller (type $funcref_=>_none) (param $x funcref)
+  ;; CHECK-NEXT:  (call_ref $A
+  ;; CHECK-NEXT:   (ref.cast $A
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $caller (export "out") (param $x funcref)
+    ;; In closed world this call must trap, as the only function of the right
+    ;; type will trap. But we are in open world so we do not optimize here.
```

Comment:
Perhaps worth a TODO about analyzing publicly exposed tables and functions to see whether it is possible for a non-trapping function to have arrived here from the outside.